gitlab-event/handler.go

```diff
@@ -0,0 +1,399 @@
+		if reqErr != nil {
+			return reqErr
+		}
+
+		digest := hmac.Sign(body, []byte(payloadSecret))
+		req.Header.Add(sdk.CloudSignatureHeader, "sha1="+hex.EncodeToString(digest))
+
+		res, err := client.Do(req)
+		if err != nil {
+			return err
+		}
+		if res.Body != nil {
+			defer res.Body.Close()
+		}
+
+		fmt.Printf("Status code returned by garbageCollect for function: `%s` by `%s` - %d\n",
+			garbageRequest.Repo, garbageRequest.Owner, res.StatusCode)
+
+		if res.StatusCode != http.StatusAccepted {
+			resBody, bodyReader := ioutil.ReadAll(res.Body)
+			if bodyReader != nil {
+				return bodyReader
+			}
+			fmt.Printf("Error in garbageCollect: %s\n", resBody)
+		}
+	}
+	return nil
+}
+
+type GarbageRequest struct {
+	Functions []string `json:"functions"`
+	Repo      string   `json:"repo"`
+	Owner     string   `json:"owner"`
+}
+
+func forward(req []byte, function string, headers map[string]string) (string, int, error) {
+	payloadSecret, err := sdk.ReadSecret("payload-secret")
+	if err != nil {
+		return "", http.StatusInternalServerError, err
+	}
+
+	suffix := os.Getenv("dns_suffix")
+	gatewayURL := os.Getenv("gateway_url")
+	gatewayURL = sdk.CreateServiceURL(gatewayURL, suffix)
+
+	c := http.Client{}
+
+	bodyReader := bytes.NewBuffer(req)
+	pushReq, reqErr := http.NewRequest(http.MethodPost, gatewayURL+"function/"+function, bodyReader)
+	if reqErr != nil {
+		return "", http.StatusBadRequest, reqErr
+	}
+	digest := hmac.Sign(req, []byte(payloadSecret))
+	pushReq.Header.Add(sdk.CloudSignatureHeader, "sha1="+hex.EncodeToString(digest))
+
+	for k, v := range headers {
+		pushReq.Header.Add(k, v)
+	}
+
+	res, err := c.Do(pushReq)
+	if err != nil {
+		msg := "cannot post to " + function + ": " + err.Error()
+		auditEvent := sdk.AuditEvent{
+			Message: msg,
+			Source:  Source,
+		}
+		sdk.PostAudit(auditEvent)
+		return "", http.StatusInternalServerError, fmt.Errorf(msg)
+	}
+
+	if res.Body != nil {
+		defer res.Body.Close()
+	}
+	body, bodyErr := ioutil.ReadAll(res.Body)
+	if bodyErr != nil {
+		return "", http.StatusInternalServerError, bodyErr
+	}
+
+	if res.StatusCode != http.StatusOK {
+		err = fmt.Errorf(string(body))
+	}
+
+	return string(body), res.StatusCode, err
+}
+
+func InstalledApp(id int, instance, apiToken string) (bool, error) {
+	wholeURL := instance + "/api/v4/projects/" + strconv.Itoa(id)
+
+	req, reqErr := http.NewRequest(http.MethodGet, wholeURL, nil)
+	if reqErr != nil {
+		return false, reqErr
+	}
+	req.Header.Add("PRIVATE-TOKEN", apiToken)
+
+	client := &http.Client{}
+	resp, respErr := client.Do(req)
+	if respErr != nil {
+		return false, respErr
+	}
+	defer resp.Body.Close()
+
+	body, bodyErr := ioutil.ReadAll(resp.Body)
+	if bodyErr != nil {
+		return false, bodyErr
+	}
+	projectInfo := GitLabProjectTags{}
+
+	unmarshalErr := json.Unmarshal(body, &projectInfo)
+	if unmarshalErr != nil {
+		return false, unmarshalErr
+	}
+
+	for _, tag := range projectInfo.TagList {
+		if strings.EqualFold(tag, "openfaas-cloud") {
```

Comment:
Can we make this tag in env var too ?
With a default to openfaas-cloud
If this solution match our needs, we probably have multiple openfaas-cloud isolated envs, and user should choose on which deploy his function
Like faas-dev and faas-prod for instance

Reply:
Very good idea! Thank you for the suggestion @WTFKr0 I will add that